File: func.js
const fdk = require("@fnproject/fdk");
const axios = require("axios");

// CAMBIAR: poner la URL del API Gateway de AWS del microservicio de pedidos
// Ejemplo: https://abc123.execute-api.us-east-1.amazonaws.com/dev
const AWS_URL = process.env.AWS_ORDERS_URL;

fdk.handle(async function(input) {

  const { clientName, address, items, phone } = input;

  if (!clientName) {
    return { status: 400, message: "Falta el nombre del cliente (clientName)" };
  }

  if (!items || items.length === 0) {
    return { status: 400, message: "El pedido no tiene productos (items)" };
  }

  const orderPayload = {
    clientName,
    address: address || "Recojo en tienda",
    phone: phone || "",
    items,
    source: "rappi",
    // CAMBIAR: poner el tenantId del restaurante que corresponda al grupo
    tenantId: "madam-tusan"
  };

  try {
    const response = await axios.post(`${AWS_URL}/orders`, orderPayload, {
      headers: { "Content-Type": "application/json" },
      timeout: 10000
    });

    return {
      status: 200,
      message: "Pedido enviado al restaurante correctamente",
      // AWS debe responder con el orderId generado
      orderId: response.data.orderId,
      estimatedTime: "30-45 minutos"
    };

  } catch (error) {
    return {
      status: 500,
      message: "No se pudo enviar el pedido al restaurante",
      detail: error.message
    };
  }

});

File: func.py
import io
import json
import os
import requests
import fdk
from fdk import response

AWS_URL = os.environ.get("AWS_ORDERS_URL", "https://vdwa84epn1.execute-api.us-east-1.amazonaws.com/dev/pedidos")

def handler(ctx, data: io.BytesIO = None):

    try:
        body = json.loads(data.getvalue())
    except Exception:
        return response.Response(
            ctx,
            response_data=json.dumps({"status": 400, "mensaje": "Body inválido"}),
            headers={"Content-Type": "application/json"}
        )

    cliente     = body.get("cliente", {})
    items       = body.get("items", [])
    monto_total = body.get("monto_total", 0.0)

    if not items:
        return response.Response(
            ctx,
            response_data=json.dumps({"status": 400, "mensaje": "Faltan items del pedido"}),
            headers={"Content-Type": "application/json"}
        )

    if not cliente.get("nombre"):
        return response.Response(
            ctx,
            response_data=json.dumps({"status": 400, "mensaje": "Falta nombre del cliente"}),
            headers={"Content-Type": "application/json"}
        )

    payload = {
        "tenant_id":   "madam-tusan",  
        "origen":      "RAPPI",        
        "cliente": {
            "nombre":    cliente.get("nombre"),
            "direccion": cliente.get("direccion", "Sin dirección"),
            "telefono":  cliente.get("telefono", "")
        },
        "items":       items,          
        "monto_total": monto_total
    }

    try:
        res = requests.post(
            f"{AWS_URL}",
            json=payload,
            headers={"Content-Type": "application/json"},
            timeout=10
        )

        data_aws = res.json()

        return response.Response(
            ctx,
            response_data=json.dumps({
                "status":    200,
                "mensaje":   "Pedido enviado al restaurante correctamente",
                "pedido_id": data_aws.get("pedido_id"),
                "tenant_id": data_aws.get("tenant_id")
            }),
            headers={"Content-Type": "application/json"}
        )

    except Exception as e:
        return response.Response(
            ctx,
            response_data=json.dumps({
                "status":  500,
                "mensaje": "Error al conectar con AWS",
                "detalle": str(e)
            }),
            headers={"Content-Type": "application/json"}
        )


if __name__ == "__main__":
    fdk.handle(handler)
